File: game/ui.py
import pygame

class UI:
    def __init__(self, width, game_height, metrics_width, bottom_height):
        self.width = width
        self.game_height = game_height
        self.metrics_width = metrics_width
        self.bottom_height = bottom_height
        
        # Colors
        self.WHITE = (255, 255, 255)
        self.BLACK = (0, 0, 0)
        self.GRAY = (128, 128, 128)
        self.BLUE = (0, 120, 255)
        
        # Font
        pygame.font.init()
        self.font = pygame.font.SysFont('Arial', 24)
        self.small_font = pygame.font.SysFont('Arial', 16)
        
        # Buttons - moved to metrics section
        pause_width = 100
        speed_width = 120  # Wider button for speed text
        button_height = 30
        button_margin = 20
        metrics_center_x = metrics_width // 2
        self.buttons = {
            'pause': pygame.Rect(
                metrics_center_x - speed_width - 10,  # Left of center, adjusted for wider speed button
                game_height + button_margin,
                pause_width,
                button_height
            ),
            'speed': pygame.Rect(
                metrics_center_x + 10,  # Right of center
                game_height + button_margin,
                speed_width,  # Using wider width for speed button
                button_height
            )
        }
        
        self.paused = False
        
    def draw(self, screen, stats):
        # Draw buttons in metrics section
        mouse_pos = pygame.mouse.get_pos()
        for button_name, button_rect in self.buttons.items():
            hover = button_rect.collidepoint(mouse_pos)
            color = (0, 150, 255) if hover else self.BLUE
            
            button_surface = pygame.Surface((button_rect.width, button_rect.height))
            button_surface.fill(color)
            button_surface.set_alpha(230)
            screen.blit(button_surface, button_rect)
            pygame.draw.rect(screen, self.WHITE, button_rect, 2)  # White outline
            
            text = "Resume" if self.paused and button_name == 'pause' else "Pause" if button_name == 'pause' else f"Speed: {stats['speed']}x"
            text_surface = self.font.render(text, True, self.WHITE)
            text_rect = text_surface.get_rect(center=button_rect.center)
            screen.blit(text_surface, text_rect)
        
        # Draw stats in metrics section
        metrics_y = self.game_height + 70  # Start below buttons
        stats_text = [
            f"Generation: {stats['generation']}",
            f"Best Fitness: {stats['best_fitness']:.0f}",
            f"Alive Birds: {stats['alive_birds']}/{stats['total_birds']}"
        ]
        
        # Add title for metrics section
        title = self.font.render("Training Metrics", True, self.WHITE)  # Changed to white
        title_rect = title.get_rect(center=(self.metrics_width//2, self.game_height + 120))  # Moved down below buttons
        screen.blit(title, title_rect)
        metrics_y = title_rect.bottom + 20  # Start metrics below title
        
        # Draw metrics in centered column
        for text in stats_text:
            text_surface = self.font.render(text, True, self.WHITE)  # Changed to white
            text_rect = text_surface.get_rect(center=(self.metrics_width//2, metrics_y))
            screen.blit(text_surface, text_rect)
            metrics_y += 40
        
    def _draw_button(self, screen, rect, text):
        # Draw button background
        pygame.draw.rect(screen, self.BLUE, rect)
        pygame.draw.rect(screen, self.BLACK, rect, 2)
        
        # Draw button text
        text_surface = self.font.render(text, True, self.WHITE)
        text_rect = text_surface.get_rect(center=rect.center)
        screen.blit(text_surface, text_rect)
        
    def _draw_text(self, screen, text, x, y):
        text_surface = self.font.render(text, True, self.BLACK)
        screen.blit(text_surface, (x, y))
        
    def handle_click(self, pos):
        """Handle mouse click on UI elements"""
        x, y = pos
        for button_name, button_rect in self.buttons.items():
            if button_rect.collidepoint(x, y):
                print(f"Button clicked: {button_name}")  # Debug print
                return button_name
        return None

File: static/js/ui.js
class UI {
    constructor(width, gameHeight, metricsWidth, bottomHeight, gameWidth) {
        this.width = width;
        this.gameHeight = gameHeight;
        this.metricsWidth = metricsWidth;
        this.bottomHeight = bottomHeight;
        this.gameWidth = gameWidth;
        this.height = gameHeight;
        
        // Colors
        this.WHITE = 'rgb(255, 255, 255)';
        this.BLACK = 'rgb(0, 0, 0)';
        this.GRAY = 'rgb(128, 128, 128)';
        this.BLUE = 'rgb(0, 120, 255)';
        this.HOVER_BLUE = 'rgb(0, 150, 255)';
        
        // Button dimensions
        this.buttonWidth = 140;
        this.speedButtonWidth = 160;
        this.buttonHeight = 40;
        this.buttonMargin = 30;
        this.buttonSpacing = 15;
        
        // Calculate button positions
        const buttonStartX = this.gameWidth + (this.metricsWidth - (this.buttonWidth * 2 + this.buttonSpacing)) / 2;
        
        // Create buttons
        this.buttons = {
            pause: {
                x: buttonStartX,
                y: 20,
                width: this.buttonWidth,
                height: this.buttonHeight
            },
            restart: {
                x: buttonStartX + this.buttonWidth + this.buttonSpacing,
                y: 20,
                width: this.buttonWidth,
                height: this.buttonHeight
            },
            speed: {
                x: this.gameWidth + (this.metricsWidth - this.speedButtonWidth) / 2,
                y: 20 + this.buttonHeight + this.buttonSpacing,
                width: this.speedButtonWidth,
                height: this.buttonHeight
            }
        };
        
        this.paused = false;
        this.font = '32px Arial';
        this.smallFont = '24px Arial';
        this.buttonAlpha = 230;
        this.mousePos = { x: 0, y: 0 };
    }
    
    handleMouseMove(event) {
        const canvasRect = event.target.getBoundingClientRect();
        const scaleX = event.target.width / canvasRect.width;
        const scaleY = event.target.height / canvasRect.height;
        
        this.mousePos = {
            x: (event.clientX - canvasRect.left) * scaleX,
            y: (event.clientY - canvasRect.top) * scaleY
        };
    }
    
    draw(ctx, stats) {
        // Draw metrics background
        ctx.fillStyle = 'rgba(0, 0, 0, 0.8)';
        ctx.fillRect(this.gameWidth, 0, this.metricsWidth, this.height);
        
        // Separator line
        ctx.strokeStyle = this.GRAY;
        ctx.lineWidth = 2;
        ctx.beginPath();
        ctx.moveTo(this.gameWidth, 0);
        ctx.lineTo(this.gameWidth, this.height);
        ctx.stroke();
        
        // Draw buttons
        for (const [buttonName, button] of Object.entries(this.buttons)) {
            const hover = this.isPointInRect(this.mousePos, button);
            const color = hover ? this.HOVER_BLUE : this.BLUE;
            
            // Button background
            ctx.fillStyle = color;
            ctx.globalAlpha = this.buttonAlpha / 255;
            ctx.fillRect(button.x, button.y, button.width, button.height);
            ctx.globalAlpha = 1;
            
            // Button border
            ctx.strokeStyle = this.WHITE;
            ctx.lineWidth = 2;
            ctx.strokeRect(button.x, button.y, button.width, button.height);
            
            // Button text
            let text;
            if (buttonName === 'pause') {
                text = this.paused ? "Resume" : "Pause";
            } else if (buttonName === 'restart') {
                text = "Restart";
            } else {
                text = `Speed: ${stats.speed}x`;
            }
            
            ctx.font = '24px Arial';
            ctx.fillStyle = this.WHITE;
            ctx.textAlign = 'center';
            ctx.textBaseline = 'middle';
            ctx.fillText(text, button.x + button.width/2, button.y + button.height/2);
        }
        
        // Draw metrics text
        let metricsY = 150;
        ctx.font = this.font;
        ctx.fillStyle = this.WHITE;
        ctx.textAlign = 'center';
        
        // Title
        const title = "Training Metrics";
        ctx.fillText(title, this.gameWidth + this.metricsWidth/2, metricsY);
        metricsY += 40;
        
        // Stats
        const statsText = [
            `Generation: ${stats.generation}`,
            `Best Fitness: ${Math.floor(stats.bestFitness)}`,
            `Alive Birds: ${stats.aliveBirds}/${stats.totalBirds}`
        ];
        
        for (const text of statsText) {
            ctx.fillText(text, this.gameWidth + this.metricsWidth/2, metricsY);
            metricsY += 35;
        }
    }
    
    isPointInRect(point, rect) {
        return point.x >= rect.x && point.x <= rect.x + rect.width &&
               point.y >= rect.y && point.y <= rect.y + rect.height;
    }
    
    handleClick(pos) {
        const canvasRect = document.getElementById('gameCanvas').getBoundingClientRect();
        const scaleX = document.getElementById('gameCanvas').width / canvasRect.width;
        const scaleY = document.getElementById('gameCanvas').height / canvasRect.height;
        
        const scaledPos = {
            x: pos.x * scaleX,
            y: pos.y * scaleY
        };
        
        for (const [buttonName, button] of Object.entries(this.buttons)) {
            if (this.isPointInRect(scaledPos, button)) {
                return buttonName;
            }
        }
        return null;
    }
} 